### ISSUE TITLE
Allow for selecting networks

### ISSUE THREAD

**Issue by Renari** [truenas/apps#743]
Currently configuration for most of the apps only consist of a port and a checkbox for host networking.

It would be nice if all networks found in `docker network ls` were available as an option instead of just host networking.

For some configurations (macvlan, ipvlan) we would need the ability to specify an IP and a default gateway and a parent interface.

**Comment by stavros-k** [comment]
That is indeed something that I do want to investigate more.

Below are some notes, with some thoughts I had when taking a brief look at that.
Adding them here as notes mainly.

---

Things that won't be able due to docker's limitation:
- `host` networking for apps that have more than 1 container. This breaks docker's dns resolution

Now assuming this:
- application with more than 1 container
- User selects a custom network
Then:
- we should create an additional network, to join the "main" application and its dependencies (eg db).

This would allow to have a stable connection between main container and db container, but still being connected to the custom network user selected.

Challenges on this:
- What happens if the custom network also includes a container that conflicts with the dns of the db?

Possible solutions:
- I think docker provides a way to assign additional dns to a container. We can probably generate a unique address for he db for example.
- The unique address should be idempotent. If it changes each time the compose file is generated, might cause issues on apps that don't react to env variables changing. (eg nextcloud and home assistant need manual change of the db address on a config file).

**Comment by Renari** [comment]
Just wanted to point out that I believe this should support inter-app communications as well, not just networks within the same stack. This is useful for things like reverse proxies, for example we could have a traefik network and join containers on that network so that traefik can act as a reverse proxy for them.
```
networks:
  traefik:
    name: traefik
    external: true
```
I don't recall off the top of my head how the DNS works in this situation, I think you need to specify a static IP when joining networks like this but as stated I'm not 100% it's been a while since I've had to do this.

**Comment by stavros-k** [comment]
Yes the custom network will be between multiple stacks, not single!

**Comment by tomerh2001** [comment]
> That is indeed something that I do want to investigate more.
Any chance there's been an update on this?

I just migrated from plain Ubuntu, and I was hoping to see some configuration show up when I install new apps (for example Jellyfin) after I install a reverse proxy (for example Traefik).

Something like:

1. Installing Traefik
2. Installing Jellyfin:
- During configuration, a "Traefik Labels" (or something similar) will show up
- I could enter the host (`traefik.http.routers.jellyfin.rule`): `jellyfin.domain.tld`
- I could enter the port (`traefik.http.services.my-service.loadbalancer.server.port`): `8086`
- Etc...

Also: This could be better presented when the stack contains multiple containers. You could potentially select the container you want to assign labels to.

My current use case for this would be:

1. Traefik
2. Homepage
3. Uptime Kuma (using [AutoKuma](https://github.com/bigboot/autokuma))

**Comment by stavros-k** [comment]
There are no plans currently for such tight integration. 
ie. prepopulated labels for traefik etc.
There are infinite number of apps that integrate with labels, and if we start adding one or two.
Then others will be requested. At some point we end up with a drop down of  hundreds of labels.

That being said,

Some apps already have label section near the bottom already.
All will get that but that's a long process as it involves updating to our v2 library on each app.

But I think you will also need the network to be added as well.
That might take a bit more.

**Comment by tomerh2001** [comment]
> There are no plans currently for such tight integration. ie. prepopulated labels for Traefik etc. There are infinite number of apps that integrate with labels, and if we start adding one or two. Then others will be requested. At some point we end up with a drop down of hundreds of labels.

I mean, It can always be a first dropdown of **Installed** applications with integrations, and only then select from a predefined list of labels **for that particular application**.

But sure, I do understand what you mean (although I might not agree that it's not a helpful feature, even if the list gets big).

Regarding generic support for key-value labels for all apps, is there a plan for it in the near future? That's kinda blocking me from having the full "Truenas experience" (as I have to use custom apps / dockge)

**Comment by stavros-k** [comment]
Each app that might integrate with labels, might have infinite number of labels. Even treafik, I can already think of hundreds of labels. Only listing them on UI will become slow at some point.

Generic key-value is already there for some apps, mainly the ones with postgres and some *arrs, plex, jelly etc.
Its a WIP, every day more apps getting those.

**Comment by Renari** [comment]
> Each app that might integrate with labels, might have infinite number of labels. Even treafik, I can already think of hundreds of labels. Only listing them on UI will become slow at some point.

Another reason this isn't possible is because many of traefiks labels aren't even static, they heavily depend on your configuration, the labels will have network names and such in them, it's an insane amount of work for such a small feature.

**Comment by tomerh2001** [comment]
> Another reason this isn't a good idea is because many of traefiks labels aren't even static, they heavily depend on your configuration, the labels will have network names and such in them, it's an insane amount of work for such a small feature.

Uh ok don't kill me guys, I was just suggesting an idea... anyway I guess it's much more important for the focus to be on supporting all of docker's configuration features than niche stuff like that :)

**Comment by voltechs** [comment]
Slightly related to this initiative would be also allowing us to set/specify the `hostname` so we could reference it (nginx proxy manager?) via the `hostname`.

**Comment by moebis** [comment]
I've been following this issue on those other closed issues and it's still not clear if there is a way to assign a different IP/NIC to this docker container? Right now the only option is to run a VM because port 53 is being used by TrueNAS for mDNS and we need port 53 in Home Assistant for Homekit bridge too. I think there might be a way to do this with a docker manifest, but I'm not sure. It works with Home Bridge docker, so I know it can be done, and in that manifest it lets you specify a NIC to use.

**Comment by stavros-k** [comment]
Library work for configuring networks per app per container is ready.
Of course after the merge slowly each app need to add support in the questions / template for it.
So it will take sometime

**Comment by Tarballwalf** [comment]
I have one question. I'm not that familiar with how TrueNAS components work. for this to be implemented, we are going to have to wait for the next major release (26.04) or is this a separate component that is be shipped/updated regardless of it?

**Comment by lumarel** [comment]
@Tarballwalf (be aware that I might have misunderstood something, so could be incorrect)
The individual apps will need to be ported to the new lib version, and of course also use the new feature set.
The catalog overall has a minimum required TrueNAS version, but there is no new version required for this change.

Theoretically you can even test it yourself with the devbox, as mentioned in the Contributing guide. (by adapting one of the apps to the new lib)

**Comment by stavros-k** [comment]
No new TrueNAS version is required.

Initial PRs are in
https://github.com/truenas/apps/pull/4382
https://github.com/truenas/apps/pull/4263

As you can understand each app needs to be individually updated. So it will take sometime.

**Comment by dough29** [comment]
Is there an already migrated app so we can check what to do ?

Thanks !

**Comment by jondycz** [comment]
Just wanted to say thanks for doing the boring and very time consuming part of porting EVERY app to allow the network selection. You're the best!

**Comment by stavros-k** [comment]
10% work, 90% cursing claude for not being able to copy paste stuff

**Comment by dough29** [comment]
Working great with qbittorrent and a macvlan network 😎 

Created a new network :

```
docker network create -d macvlan \
  --subnet=192.168.123.0/24 \
  --gateway=192.168.123.1 \
  -o parent=br0 \
  br0_macvlan
```

In the app setup, network name to _br0_macvlan_ and IPv4 Address to _192.168.123.10_

For port exposition, use option _Publish port on the host for external access_, it will be exposed directly on the selected IP : ex. _192.168.123.10:8080_

**Comment by VladFlorinIlie** [comment]
Is there any way to create the docker network from TrueNas UI/middleware directly? Or at least store the network creation in the TrueNas config?

**Comment by jondycz** [comment]
> Is there any way to create the docker network from TrueNas UI/middleware directly? Or at least store the network creation in the TrueNas config?

Yes, in truenas network settings. The same way you created your bridge.

**Comment by VladFlorinIlie** [comment]
> > Is there any way to create the docker network from TrueNas UI/middleware directly? Or at least store the network creation in the TrueNas config?
> 
> Yes, in truenas network settings. The same way you created your bridge.

I don't see any option in truenas network settings for creating a docker network. I cannot add another bridge and there is no "docker network" interface if that's what you meant. Is this documented somewhere?

**Comment by jondycz** [comment]
> > > Is there any way to create the docker network from TrueNas UI/middleware directly? Or at least store the network creation in the TrueNas config?
> > 
> > 
> > Yes, in truenas network settings. The same way you created your bridge.
> 
> I don't see any option in truenas network settings for creating a docker network. I cannot add another bridge and there is no "docker network" interface if that's what you meant. Is this documented somewhere?

IIRC, "docker network" is just a network that containers join. It's not specific to docker. You create a network in trueNAS (Settings->Network) and then use the name of that network in the app of your choice so the app joins the network.

**Comment by VladFlorinIlie** [comment]
> > > > Is there any way to create the docker network from TrueNas UI/middleware directly? Or at least store the network creation in the TrueNas config?
> > > 
> > > 
> > > Yes, in truenas network settings. The same way you created your bridge.
> > 
> > 
> > I don't see any option in truenas network settings for creating a docker network. I cannot add another bridge and there is no "docker network" interface if that's what you meant. Is this documented somewhere?
> 
> IIRC, "docker network" is just a network that containers join. It's not specific to docker. You create a network in trueNAS (Settings->Network) and then use the name of that network in the app of your choice so the app joins the network.

Unfortunately, no. A docker network is managed entirely by the docker engine, and truenas network settings does not know (at least not through GUI) how to create such network.

So it seems that is not possible to do what I was thinking of. Maybe this is a request specifically for a new version of truenas?
If anyone has any other clues regarding this, please correct me. I see this automatic docker network creation / storage in truenas config as a necessary step for restoring truenas from backups or when upgrading. AFAIK, commands done through CLI (like docker network) do not survive truenas upgrades.

**Comment by dough29** [comment]
Here I'm not able to set the network name for one already existing in the UI, I get "has to be created before" error.

That's why I created the docker macvlan network.

**Comment by VladFlorinIlie** [comment]
> Here I'm not able to set the network name for one already existing in the UI, I get "has to be created before" error.
> 
> That's why I created the docker macvlan network.

Indeed, I have the same error: " Network [name] has to be created before it can be used. ie via Docker CLI". But as I said, from what I know, CLI commands do not survive truenas upgrades. If this is true, at the next upgrade, all the docker network setup in apps configurations will need to be recreated.

**Comment by dough29** [comment]
I think that what you set in Docker should survive any reboot/update.

**Comment by stavros-k** [comment]
Docker networks currently have to be created via cli or other apps like portainer. Reboots and upgrades won't wipe those. It persists as part of the docker data. Later will be able to do that via ui, but this is not implemented yet

**Comment by sylveon** [comment]
Once this functionality is implemented in UI, would it conflict with existing manually-created networks? Is there a tracking issue for networks?

**Comment by aserrallerios** [comment]
Can anybody confirm that this improvement does indeed fix https://github.com/truenas/apps/issues/583?

Should I just expose a port of the HA container in the host network and this will autodiscover devices in the same home network?

**Comment by silicon359** [comment]
@aserrallerios You cannot just expose a port of the HA app and autodiscover devices on your home network. That just grants access to the HA web interface. At that point, network scanning will only scan the Docker network. You will need to do what @dough29 said here: https://github.com/truenas/apps/issues/743#issuecomment-3936347907 but edited with your network info.

Mine looked like this

```
sudo docker network create -d macvlan \
  --subnet=192.168.0.0/24 \
  --gateway=192.168.0.1 \
  --ip-range=192.168.0.254/32 \
  -o parent=enp1s0 \
  br1_macvlan
```

That created the macvlan bridge with a subnet that contained one static IP on my network (outside DHCP). Then I added a network in the app GUI that connected the HA container to that macvlan, and assigned the static IP. That worked for me and did not break communication with postgres in its own container and HA was able to communicate with things on the regular network.

IPs changed to protect the innocent.

**Comment by aserrallerios** [comment]
Yeah thanks @silicon359, validated that macvlan network worked perfectly and allowed HA to scan devices in the home network.
